docs/docs/recipes/single-select-a11y-server-side-filtering.js

```diff
@@ -0,0 +1,260 @@
+import { useDataEngine, useDataQuery } from '@dhis2/app-runtime'
+import { SingleSelectA11y } from '@dhis2/ui'
+import React, { useCallback, useEffect, useRef, useState } from 'react'
+import { CustomDataProvider } from './single-select-a11y-server-side-filtering/index.js'
+
+function dataElementToOption({ id, displayName }) {
+    return { value: id, label: displayName }
+}
+
+function useLoadDataElementQuery(options) {
+    return useDataQuery(
+        {
+            result: {
+                resource: 'dataElements',
+                id: ({ id }) => id,
+            },
+        },
+        { ...options, lazy: true }
+    )
+}
+
+function useLoadDataElementsQuery(options) {
+    return useDataQuery(
+        {
+            result: {
+                resource: 'dataElements',
+                params: ({ page }) => ({ page, pageSize: 10 }),
+            },
+        },
+        options
+    )
+}
+
+function useLoadFilteredDataElementsQuery(customOptions) {
+    const engine = useDataEngine()
+    const timeout = useRef(null)
+    const abortController = useRef()
+    const [loading, setLoading] = useState(false)
+    const [error, setError] = useState(null)
+
+    const refetch = useCallback(
+        ({ searchTerm, page }) => {
+            if (timeout.current) {
```

Comment:
I feel like this is unnecessarily complex? Do I understand correctly that this is to prevent refetches happening when one is already in flight, potentially overwriting data? Why don't we just check `isLoading` before calling `refetch`?

Reply:
> Why don't we just check isLoading before calling refetch

In that case you'd have to wait for the result after ever character you enter into the search field.
When scrolling down the list, the same `refetch` function is used, but that's just to keep things DRY.

> I would also think this could potentially "skip" a page if a next-page is inflight and it's called again within 200ms, cancelling the previous page?

Loading the next page is actually prevented when `loading` is true (see [here](https://github.com/dhis2/ui/blob/98514fd7fc99ae2e586a7f1a3cec493e9c5bdd56/docs/docs/recipes/single-select-a11y-server-side-filtering.js#L200)).
It will cancel the previous request when loading the next page and then changing the filter while the request is still outgoing.

Reply:
> It's very thorough tho - I will give you that!

The idea was to show the simplest filtering as well as the most complex, allowing the user to either add a little (to the simple use case) or remove from the thorough examples.

Adding more examples is always a good idea IMO when it makes sense!
Should I add one like this: Always load the same amount of items when not searching (showing the latest values only) and displaying all search results, unpaginated)